### ISSUE TITLE
Allow different shapes for `y_true` and `y_pred` in graph mode

### ISSUE THREAD

**Issue by tom-andersson** [tensorflow/tensorflow#48736]
**System information**
- TensorFlow version (you are using): 2.2.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
Currently, running Keras in graph mode means that `y_true` and `y_pred` in the `loss` function passed to `Model.fit()` must have the same shape. Although this works in eager mode, I would like to be able to train models where `y_true` and `y_pred` are different shapes while benefiting from graph mode. My use case is that `y_true` has an extra final dimension of size 2, where index 0 gives the labels and index 1 gives a sample weight which I use for weighting my loss function (e.g. multiplying some pixels by zero).

This relates to the following issue in the Keras GitHub but there wasn't a satisfying solution provided: https://github.com/keras-team/keras/issues/4781

For the time being I'm simply adding an extra dimension of size 1 to the `y_pred` from my model using `tf.expand_dims`. However, this is inconvenient for the rest of my codebase where I now have to slice out the final dimension of my model's predictions.

Apologies if this has already been addressed in a later version of TF or if I've missed something...

**Comment by deeb02** [comment]
Keras supports sample weighting in a separate argument, why does that not cover your use case?

**Comment by tom-andersson** [comment]
@deeb02 Thank you for your reply. Does the passing `sample_weight` as a third element from a data loader in `Model.fit` work with custom metrics?

**Comment by laplacericky** [comment]
@tom-andersson Yes, it should work with custom metrics or loss but your custom metrics or loss class must support sample weights
